### PR TITLE
tfm: Add zephyr native ns_interface and logging function

### PR DESCRIPTION
**Pull request by oyvindronningstad**
To allow using TFM NS interface without enabling CMSIS_RTOS support.
And to allow using TFM ns code that uses logging.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Review by microbuilder — APPROVED** [review]

File: modules/trusted-firmware-m/CMakeLists.txt [hunk 1]
```diff
@@ -256,6 +256,8 @@ if (CONFIG_BUILD_WITH_TFM)
   )
 
   zephyr_link_libraries(tfm_api)
+  zephyr_sources(
+    ${ZEPHYR_BASE}/modules/trusted-firmware-m/src/zephyr_tfm_ns_interface.c)
```
Comment (tejlmand) [inline modules/trusted-firmware-m/CMakeLists.txt:260]: Why is this a `zephyr_sources` ?
Does it really belong in the `zephyr` lib.

**Comment by oyvindronningstad** [comment]
Merged as part of #33510.

**Comment by tejlmand** [comment]
> Merged as part of #33510.

Seems that nullified my comment :laughing:

**Comment by ioannisg** [comment]
> > Merged as part of #33510.
> 
> Seems that nullified my comment 😆

I think this source is part of zephyr lib, @tejlmand. But I guess @oyvindronningstad could still answer, regardless of the change being already merged. :)

**Comment by oyvindronningstad** [comment]
I just wanted to stop/redirect more comments on this. I will answer your question @tejlmand. It's probably "that's what worked after a couple of hours struggling a to get cmake to give me my symbols" 🙃

File: modules/trusted-firmware-m/CMakeLists.txt (hunk 1 above)
Comment (oyvindronningstad) [reply modules/trusted-firmware-m/CMakeLists.txt:260]: It is part of the nonsecure image, to help interfacing with tfm.

**Comment by tejlmand** [comment]
> I think this source is part of zephyr lib, @tejlmand. But I guess @oyvindronningstad could still answer, regardless of the change being already merged. :)

I took a second look.

Actually, you don't know which library the file will end up in.

At first glance, this looks reasonable, as it will be added to the currently defined library (usually not `libzephyr.a`)
https://github.com/zephyrproject-rtos/zephyr/blob/fed035231fe9d4fc9d63e7367b4fa5a658861daf/modules/trusted-firmware-m/CMakeLists.txt#L208-L210

until one realizes there is NO `zephyr_library()` call in this CMakeLists.txt.
Because this is included as `ZEPHYR_<MODULE>_CMAKE_DIR` and `zephyr` is the current library, then you get the code in the `zephyr` lib, but if there had been another `zephyr_library()` call prior to this code in the parent scope, then the code would have ended up inside that particular library.

When adding sources to `zephyr` lib, then `zephyr_sources` should be used and not `zephyr_library_sources`.
Also, there should be a good reason for adding sources to `zephyr` lib, and not just *then it works*.
See this comment, on how `zephyr` lib may look:
https://github.com/zephyrproject-rtos/zephyr/issues/8825#issue-339723582
 

> It's probably "that's what worked after a couple of hours struggling a to get cmake to give me my symbols" upside_down_face

Sound like a bad approach.

**Pull request closed without merging**